opentelemetry.go

```diff
@@ -0,0 +1,203 @@
+package amqp091
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/getoutreach/gobox/pkg/app"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/propagation"
+	semconv "go.opentelemetry.io/otel/semconv/v1.25.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// tracer is the tracer used by the package
+var tracer = otel.Tracer("amqp091")
+
+// amqpHeaderCarrier is a carrier for AMQP headers.
+type amqpHeaderCarrier Table
+
+// Get returns the value associated with the passed key.
+func (c amqpHeaderCarrier) Get(key string) string {
+	v, ok := c[key]
+	if !ok {
+		return ""
+	}
+	s, ok := v.(string)
+	if ok {
+		return s
+	}
+	return ""
+}
+
+// Set stores the key-value pair.
+func (c amqpHeaderCarrier) Set(key, value string) {
+	c[key] = value
+}
+
+// Keys lists the keys stored in this carrier.
+func (c amqpHeaderCarrier) Keys() []string {
+	keys := []string{}
+	for k, v := range c {
+		if _, ok := v.(string); !ok {
+			continue
+		}
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+// ensure amqpHeaderCarrier implements the TextMapCarrier interface
+var _ propagation.TextMapCarrier = amqpHeaderCarrier{}
+
+// InjectSpan injects the span context into the AMQP headers.
+// It returns the input headers with the span headers added.
+func injectSpanFromContext(ctx context.Context, headers Table) Table {
+	carrier := amqpHeaderCarrier(headers)
+	if carrier == nil {
+		carrier = amqpHeaderCarrier{}
+	}
+	otel.GetTextMapPropagator().Inject(ctx, carrier)
+	return Table(carrier)
+}
+
+// ExtractSpanContext extracts the span context from the AMQP headers.
+func ExtractSpanContext(ctx context.Context, headers Table) context.Context {
+	carrier := amqpHeaderCarrier(headers)
+	if carrier == nil {
+		carrier = amqpHeaderCarrier{}
+	}
+	return otel.GetTextMapPropagator().Extract(ctx, carrier)
+}
+
+// extractSpanFromReturn creates a span for a returned message
+func extractSpanFromReturn(
```

Comment:
ahh I haven't wired the return up yet. Probably gets a similar treatment to Delivery, if that works.

Reply:
The [rabbitmq semconv specs](https://opentelemetry.io/docs/specs/semconv/messaging/rabbitmq/) does not mention how message returns should be instrumented, maybe we should open an issue in https://github.com/open-telemetry/semantic-conventions/issues asking for clarification.